File: src/services/weatherService.ts
import { OpenMeteoWeatherResponse } from '@/types/openMeteo';
import { FASTAPI_BASE_URL } from '@/config/api';

const OPEN_METEO_WEATHER_BASE_URL = 'https://api.open-meteo.com/v1/forecast';


/**
 * Fetch weather from ORCA FastAPI backend (http://localhost:8000/api/weather)
 */
export async function fetchFastAPIWeather(lat: number, lon: number) {
  if (typeof lat !== 'number' || typeof lon !== 'number' || isNaN(lat) || isNaN(lon)) {
    throw new Error(`Invalid coordinates passed to fetchFastAPIWeather: lat=${lat}, lon=${lon}`);
  }
  const response = await fetch(`${FASTAPI_BASE_URL}/api/weather?latitude=${lat}&longitude=${lon}`, {
    cache: 'no-store',
  });
  if (!response.ok) {
    throw new Error(`FastAPI /api/weather returned HTTP ${response.status}`);
  }
  return response.json();
}

/**
 * Fetch real weather data from Open-Meteo Forecast API
 * No API key required.
 */
export async function fetchWeatherForecast(lat: number, lon: number): Promise<OpenMeteoWeatherResponse> {
  if (typeof lat !== 'number' || typeof lon !== 'number' || isNaN(lat) || isNaN(lon)) {
    throw new Error(`Invalid coordinates passed to fetchWeatherForecast: lat=${lat}, lon=${lon}`);
  }
  const params = new URLSearchParams({
    latitude: lat.toString(),
    longitude: lon.toString(),
    current: [
      'temperature_2m',
      'relative_humidity_2m',
      'precipitation',
      'wind_speed_10m',
      'wind_direction_10m',
      'wind_gusts_10m',
    ].join(','),
    hourly: [
      'temperature_2m',
      'relative_humidity_2m',
      'precipitation',
      'wind_speed_10m',
      'wind_direction_10m',
      'wind_gusts_10m',
    ].join(','),
    wind_speed_unit: 'kn',
    timezone: 'Asia/Kolkata',
  });

  const response = await fetch(`${OPEN_METEO_WEATHER_BASE_URL}?${params.toString()}`, {
    cache: 'no-store',
  });

  if (!response.ok) {
    throw new Error(`Open-Meteo Weather API responded with HTTP ${response.status}: ${response.statusText}`);
  }

  return response.json();
}

File: backend/services/weather_service.py
import httpx
from typing import Dict, Any

OPEN_METEO_WEATHER_URL = "https://api.open-meteo.com/v1/forecast"

def degrees_to_compass(deg: float) -> str:
    """Convert degrees to 8-point compass bearing."""
    directions = ["N", "NE", "E", "SE", "S", "SW", "W", "NW"]
    idx = round(((deg % 360) + 360) % 360 / 45) % 8
    return directions[idx]

async def get_weather_data(lat: float, lon: float) -> Dict[str, Any]:
    """
    Retrieve real atmospheric weather data from Open-Meteo Forecast API.
    Requires explicit latitude and longitude coordinates.
    No API key required.
    """
    params = {
        "latitude": lat,
        "longitude": lon,
        "current": [
            "temperature_2m",
            "relative_humidity_2m",
            "precipitation",
            "wind_speed_10m",
            "wind_direction_10m",
            "wind_gusts_10m",
        ],
        "hourly": [
            "temperature_2m",
            "relative_humidity_2m",
            "precipitation",
            "wind_speed_10m",
            "wind_direction_10m",
            "wind_gusts_10m",
        ],
        "wind_speed_unit": "kn",
        "timezone": "Asia/Kolkata",
    }

    async with httpx.AsyncClient(timeout=10.0) as client:
        response = await client.get(OPEN_METEO_WEATHER_URL, params=params)
        response.raise_for_status()
        raw = response.json()

    current = raw.get("current", {})
    hourly = raw.get("hourly", {})
    wind_dir = current.get("wind_direction_10m", 0.0)

    return {
        "source": "Open-Meteo",
        "coordinates": {"latitude": lat, "longitude": lon},
        "current": {
            "temperature": current.get("temperature_2m"),
            "temperature_unit": "°C",
            "humidity": current.get("relative_humidity_2m"),
            "humidity_unit": "%",
            "precipitation": current.get("precipitation"),
            "precipitation_unit": "mm",
            "wind_speed": current.get("wind_speed_10m"),
            "wind_speed_unit": "kn",
            "wind_direction": wind_dir,
            "wind_direction_compass": degrees_to_compass(wind_dir),
            "wind_gusts": current.get("wind_gusts_10m"),
            "wind_gusts_unit": "kn",
            "time": current.get("time"),
        },
        "hourly": {
            "time": hourly.get("time", [])[:48],
            "temperature": hourly.get("temperature_2m", [])[:48],
            "precipitation": hourly.get("precipitation", [])[:48],
            "wind_speed": hourly.get("wind_speed_10m", [])[:48],
            "wind_direction": hourly.get("wind_direction_10m", [])[:48],
            "wind_gusts": hourly.get("wind_gusts_10m", [])[:48],
            "humidity": hourly.get("relative_humidity_2m", [])[:48],
        },
    }
